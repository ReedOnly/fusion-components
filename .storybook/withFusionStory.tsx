import React from 'react';
import Header from '../src/components/core/Header';
import Content from '../src/components/core/Content';
import Overlay from '../src/components/general/ApplicationGuidance/components/Overlay';

export default (title: string, padding: number = 16) => (stories: () => React.ReactNode) => {
    return (
        <>
            <Overlay fixed>
                <Header aside={null} start={null} content={null} />
            </Overlay>
            <Content>
                <Overlay>
                    <div
                        style={{
                            padding,
                            minHeight: '100%',
                            width: '100%',
                            maxWidth: '100%',
                            boxSizing: 'border-box',
                        }}
                    >
                        {stories()}
                    </div>
                </Overlay>
            </Content>
        </>
    );
};
