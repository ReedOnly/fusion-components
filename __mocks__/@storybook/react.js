import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    // const decoratedStory = api.decorators.reduce((s, decorator) => {
                    //     try {
                    //         return decorator(s, { parameters: api.parameters });
                    //     } catch (e) {
                    //         return s;
                    //     }
                    // }, story);

                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
