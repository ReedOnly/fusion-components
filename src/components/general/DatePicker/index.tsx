import React, { useState, useMemo, useCallback, useEffect } from 'react';
import { dateMask, formatDate, parseDate, Month } from '@equinor/fusion';
import {
    Dropdown,
    useDropdownController,
    TextInput,
    useStringMask,
    unmaskString,
    Calendar,
    CalendarIcon,
} from '@equinor/fusion-components';

type DatePickerProps = {
    label?: string;
    selectedDate?: Date | null;
    onChange: (date: Date | null) => void;
};

const DatePicker: React.FC<DatePickerProps> = ({ label, selectedDate, onChange }) => {
    const [inputValue, setInputValue] = useState<string>('');
    const [maskedValue, isValidMask] = useStringMask(dateMask, inputValue);

    const dropdownController = useDropdownController((ref, isOpen, setIsOpen) => {
        const value = useMemo(() => {
            if (!selectedDate || isOpen) {
                return maskedValue;
            }

            return formatDate(selectedDate);
        }, [selectedDate, isOpen, maskedValue]);

        const handleKeyUp = useCallback((e: React.KeyboardEvent<HTMLInputElement>) => {
            if(e.keyCode === 13 && isValidMask) {
                try {
                    const parsedDate = parseDate(maskedValue);
                    onChange(parsedDate);
                    setIsOpen(false);
                } catch (e) {}
            } else if(e.keyCode === 27) {
                setIsOpen(false);
            }
        }, [isValidMask, onChange]);

        return (
            <TextInput
                value={value}
                onChange={value => setInputValue(unmaskString(dateMask, value))}
                onIconAction={() => isOpen && setIsOpen(false)}
                onClick={() => !isOpen && setIsOpen(true)}
                icon={<CalendarIcon />}
                label={label}
                placeholder={selectedDate ? formatDate(selectedDate) : 'dd/mm/yyyy'}
                onKeyUp={handleKeyUp}
            />
        );
    });

    useEffect(() => {
        if(!dropdownController.isOpen) {
            setInputValue('');
        }
    }, [dropdownController.isOpen]);

    const onDatePickerChange = useCallback(
        (date: Date) => {
            onChange(date);
            dropdownController.setIsOpen(false);
        },
        [onChange, dropdownController.setIsOpen]
    );

    const today = new Date();

    return (
        <div
            ref={dropdownController.controllerRef as React.MutableRefObject<HTMLDivElement | null>}
        >
            <Dropdown controller={dropdownController}>
                <Calendar
                    selectedDate={selectedDate}
                    initialYear={today.getFullYear()}
                    initialMonth={today.getMonth() as Month}
                    interactive
                    onChange={onDatePickerChange}
                />
            </Dropdown>
        </div>
    );
};

export default DatePicker;
