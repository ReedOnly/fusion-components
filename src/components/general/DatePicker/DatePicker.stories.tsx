import * as React from 'react';
import { storiesOf } from '@storybook/react';
import DatePicker from './index';
import withFusionStory from '../../../../.storybook/withFusionStory';
import { formatDate } from '@equinor/fusion';

const DatePickerStory = () => {
    const [selectedDate, setSelecteDate] = React.useState<Date | null>(null);

    return (
        <div style={{ width: 200 }}>
            <DatePicker selectedDate={selectedDate} onChange={setSelecteDate} />
            Selected date: {selectedDate ? formatDate(selectedDate) : null}
        </div>
    );
};

storiesOf('General|Date picker', module)
    .addDecorator(withFusionStory('Date picker'))
    .add('Default', () => <DatePickerStory />);
