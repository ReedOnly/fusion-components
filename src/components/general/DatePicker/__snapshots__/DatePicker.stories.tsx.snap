// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General|Date picker Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      style={
        Object {
          "width": 200,
        }
      }
    >
      <div>
        <div
          className="inputContainer"
        >
          <div
            className="inputContent"
            onClick={[Function]}
          >
            <div
              className="inputTextContent"
            >
              <label>
                Select date
              </label>
              <input
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyUp={[Function]}
                placeholder=""
                value=""
              />
            </div>
            <div
              className="icon"
              onClick={[Function]}
            >
              <svg
                height={24}
                style={
                  Object {
                    "color": "",
                  }
                }
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  clipRule="evenodd"
                  d="M19 4H18V3C18 2.45 17.55 2 17 2C16.45 2 16 2.45 16 3V4H8V3C8 2.45 7.55 2 7 2C6.45 2 6 2.45 6 3V4H5C3.89 4 3.01 4.9 3.01 6L3 20C3 21.1 3.89 22 5 22H19C20.1 22 21 21.1 21 20V6C21 4.9 20.1 4 19 4ZM19 19C19 19.55 18.55 20 18 20H6C5.45 20 5 19.55 5 19V9H19V19ZM9 13V11H7V13H9ZM11 11H13V13H11V11ZM17 13V11H15V13H17Z"
                  fill="currentColor"
                  fillRule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <br />
      <div>
        <div
          className="inputContainer"
        >
          <div
            className="inputContent error"
            onClick={[Function]}
          >
            <div
              className="inputTextContent error"
            >
              <label>
                Select date
              </label>
              <input
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyUp={[Function]}
                placeholder=""
                value=""
              />
            </div>
            <div
              className="icon"
              onClick={[Function]}
            >
              <svg
                height={24}
                style={
                  Object {
                    "color": "var(--color-red)",
                  }
                }
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  clipRule="evenodd"
                  d="M12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2ZM12 13C11.45 13 11 12.55 11 12V8C11 7.45 11.45 7 12 7C12.55 7 13 7.45 13 8V12C13 12.55 12.55 13 12 13ZM11 15V17H13V15H11Z"
                  fill="currentColor"
                  fillRule="evenodd"
                />
              </svg>
            </div>
          </div>
          <div
            className="helperText error"
          >
            Select a date in the future
          </div>
        </div>
      </div>
      <br />
      <div>
        <div
          className="inputContainer"
        >
          <div
            className="inputContent disabled"
            onClick={[Function]}
          >
            <div
              className="inputTextContent disabled"
            >
              <label>
                Select date
              </label>
              <input
                disabled={true}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyUp={[Function]}
                placeholder=""
                value=""
              />
            </div>
            <div
              className="icon"
              onClick={[Function]}
            >
              <svg
                height={24}
                style={
                  Object {
                    "color": "",
                  }
                }
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  clipRule="evenodd"
                  d="M19 4H18V3C18 2.45 17.55 2 17 2C16.45 2 16 2.45 16 3V4H8V3C8 2.45 7.55 2 7 2C6.45 2 6 2.45 6 3V4H5C3.89 4 3.01 4.9 3.01 6L3 20C3 21.1 3.89 22 5 22H19C20.1 22 21 21.1 21 20V6C21 4.9 20.1 4 19 4ZM19 19C19 19.55 18.55 20 18 20H6C5.45 20 5 19.55 5 19V9H19V19ZM9 13V11H7V13H9ZM11 11H13V13H11V11ZM17 13V11H15V13H17Z"
                  fill="currentColor"
                  fillRule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
